sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetReadSupport.scala

```diff
@@ -323,15 +344,27 @@ object ParquetReadSupport {
    * @return A list of clipped [[GroupType]] fields, which can be empty.
    */
   private def clipParquetGroupFields(
-      parquetRecord: GroupType, structType: StructType, caseSensitive: Boolean): Seq[Type] = {
+      parquetRecord: GroupType,
+      structType: StructType,
+      caseSensitive: Boolean,
+      accessByColumnOrdinal: Boolean): Seq[Type] = {
     val toParquet = new SparkToParquetSchemaConverter(writeLegacyParquetFormat = false)
-    if (caseSensitive) {
+    if (accessByColumnOrdinal) {
+      structType.zipWithIndex.map { case (f, i) =>
+        if (i < parquetRecord.getFieldCount) {
+          clipParquetType(
+            parquetRecord.getType(i), f.dataType, caseSensitive, accessByColumnOrdinal)
+        } else {
+          toParquet.convertField(f)
+        }
+      }
+    } else if (caseSensitive) {
       val caseSensitiveParquetFieldMap =
         parquetRecord.getFields.asScala.map(f => f.getName -> f).toMap
       structType.map { f =>
         caseSensitiveParquetFieldMap
           .get(f.name)
-          .map(clipParquetType(_, f.dataType, caseSensitive))
+          .map(clipParquetType(_, f.dataType, caseSensitive, accessByColumnOrdinal))
```

Comment:
why do we need to pass down `accessByColumnOrdinal`? it is always false here right? also it seems we can't do this on nested columns?

Reply:
both accessByColumnOrdinal or `false` are the same

Reply:
we support nesting, as you can see from the newly added tests